keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,630 @@
+
+## Proposal
+
+This section describes each of the goals in the [Goals](#goals) section in more detail.
+
+### Object selector
+
+Currently Admission Webhook supports namespace selectors, but that may not be enough for some cases that admission webhook need to be skipped on some objects. For example if the Admission Webhook is running inside cluster and its rules includes wildcards which match required API objects for its own execution, it won't work. An object selector would be useful exclude those objects. 
+
+Also in case of an optional webhooks, an object selector gives the end user ability to include or exclude an object without having access to admission webhook configuration which is normally restricted to cluster admins.
+
+Note that namespaced objects must match both the object selector (if specified) and namespace selector to be sent to the Webhook. 
+
+The object selector applies to an Object's labels. For create and update, the object is the new object. For delete, it is existing object (passed as `oldObject` in the `admissionRequest`). For sub-resources, if the subresource accepts the whole object and apply the changes to object's labels (e.g. `pods/status`) the labels on the new object will be used otherwise (e.g. `pods/proxy`) the labels on the existing parent object will be used. If subresource does not have a parent object, the Webhook call may fail or skipped based on the failure policy.
+
+The propose change is to add an ObjectSelector to the webhook API both in v1 and v1beta1:
+
+```golang
+type Webhook struct {
+    ...
+     // ObjectSelector decides whether to run the webhook on an object based
+     // on whether the object.metadata.labels matches the selector. A namespace object must
+     // match both namespaceSelector (if present) and objectSelector (if present).
+     // For sub-resources, if the labels on the newObject is effective (e.g. pods/status)
+     // they will be used, otherwise the labels on the parent object will be used. If the
+     // sub-resource does not have a parent object, the call may fail or skipped based on
+     // the failure policy.
+     //
+     // Default to the empty LabelSelector, which matches everything.
+     // +optional
+     ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,7,opt,name=objectSelector"`
+}
+```
+
+### Scope
+
+Current webhook Rules applies to objects of all scopes. That means a Rule can uses wildcards to target both namespaced and cluster scoped objects. The proposal is to add a scope field to Admission Webhook configuration to limit webhook target on namespaced object or cluster scoped objects. This enables webhook developers to target all namespace objects or all cluster-scoped objects. The field will be added to both v1 and v1beta1. The field is optional and empty value means no scope restriction.
```

Comment:
Can a delete collection call be issued for a namespaced resource at the cluster scope? (e.g. does `DELETE /api/v1/pods` work)?

I sort of hope not, because the questions it raises are difficult. 

If it does:
* would that request match a namespace scoped admission plugin because the resources were namespaced, or a cluster scoped plugin because the request was cluster scoped?
* How would namespaceSelector interact with that request? Would you only get called with the items in the list inside namespaces matching the selector?

Reply:
answering my own question: no, deletecollection only works inside a namespace for namespaced resources

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L456-L463

Reply:
mostly just calling out the implication of moving forward with this for reviewers (I'm still in favor of this)